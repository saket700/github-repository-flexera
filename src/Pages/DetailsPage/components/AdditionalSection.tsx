import React from "react";

interface IAdditionalSection {
  HeadingText: string;
  DescriptionText: string;
}
export const AdditionalSection: React.FC<IAdditionalSection> = ({
  HeadingText,
  DescriptionText,
}) => {
  return (
    <div>
      <h5 className="mb-2 text-xl font-bold tracking-tight text-gray-900 flex items-center">
        {HeadingText}
      </h5>
      <p className="font-normal text-gray-700 dark:text-gray-400">
        {DescriptionText}
      </p>
    </div>
  );
};
