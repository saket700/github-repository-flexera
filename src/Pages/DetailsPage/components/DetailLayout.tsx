import React, { ReactElement } from "react";
import {
  ReadMeIcon,
  StarIcon,
  LicenseIcon,
  ActivityIcon,
  ForkIcon,
  EyeIcon
} from "../../../Icons";
import { ListItem } from "./ListItem";

interface IDetailsLayout {
  description: string | ReactElement;
  stargazersCount: number;
  watchersCount: number;
  forksCount: number;
}

export const DetailsLayout: React.FC<IDetailsLayout> = ({
  description,
  stargazersCount,
  watchersCount,
  forksCount,
}) => {
  
  return (
    <div className="space-y-8">
      <h4 className="flex items-center text-lg font-bold dark:text-black">
        About
      </h4>
      <div className="flex items-center">{description}</div>
      <ul className="max-w-md list-inside dark:text-gray-400">
        <ListItem icon={<ReadMeIcon />} children="Readme" />
        <ListItem icon={<LicenseIcon />} children="Unlicense license" />
        <ListItem icon={<ActivityIcon />} children="Activity" />
        <ListItem icon={<StarIcon />} children={`${stargazersCount} stars`} />
        <ListItem icon={<EyeIcon />} children={`${watchersCount} watching`} />
        <ListItem icon={<ForkIcon />} children={`${forksCount} forks`} />
      </ul>
    </div>
  );
};

