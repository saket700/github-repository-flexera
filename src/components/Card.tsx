import React from "react";
import { GithubIcon, StarIcon, LicenseIcon, ForkIcon } from "../Icons";

export interface ICardDetails {
  githubUsername: string;
  githubDescription: string | undefined;
  numberOfStars?: string;
  githubLanguage: string;
  visibility: string;
  stargazersCount: number;
  updatedAt: string;
  repoName: string;
  license: {
    key: string;
    name: string;
  };
  forkCount: number;
  avatarUrl: string;
  onStarClick?: () => void;
  isStarred: boolean
}

export const Card: React.FC<ICardDetails> = ({
  githubUsername,
  githubDescription,
  githubLanguage,
  visibility,
  stargazersCount,
  updatedAt,
  repoName,
  license,
  forkCount,
  avatarUrl,
  onStarClick,
  isStarred
}) => {
  return (
    <div className="max-w-[52rem] w-full p-6 bg-white border border-gray-200 rounded-lg shadow flex flex-col md:flex-row md:items-center md:space-x-4">
      <img
        src={avatarUrl}
        alt={`${githubUsername}'s Avatar`}
        className="rounded-full w-36 h-36 mb-4 md:mb-0"
      />
      <div className="flex flex-col flex-grow">
        <div className="flex-star-container flex">
          <div className="flex items-center  mb-3 flex-1 ">
            <GithubIcon />
            <a
              href={`/detail?username=${githubUsername}&reponame=${repoName}`}
              target="_blank"
              className="ml-2 text-blue-500 hover:underline"
            >
              {repoName}
            </a>
            <div className="border text-sm text-gray-500 rounded-3xl w-16 h-[24px] ml-2 text-uppercase capitalize">
              {visibility}
            </div>
          </div>
          <div className="star cursor-pointer" onClick={onStarClick}>
            <StarIcon fill = {isStarred ? "#e3b341" : "#000;"} />
          </div>
        </div>
        <div className="flex items-start flex-wrap">
          <p className="mb-3 font-normal text-gray-500 dark:text-gray-400 text-sm text-justify">
            {githubDescription ? githubDescription : "-"}
          </p>
        </div>
        <div className="flex pt-2 text-xs flex-wrap">
          {!!stargazersCount && (
            <>
              <p className="mb-3 font-normal text-gray-500 dark:text-gray-400 mr-3">
                {githubLanguage}
              </p>
              <StarIcon />
              <div className="font-normal text-gray-400">{stargazersCount}</div>
            </>
          )}
          {!!forkCount && (
            <div className="flex pl-2">
              <ForkIcon />
              <div className="font-normal text-gray-500 dark:text-gray-400 mr-3">
                {forkCount}
              </div>
            </div>
          )}
          {!!license?.name && (
            <div className="flex pl-2">
              <LicenseIcon />
              <div className="font-normal text-gray-500 dark:text-gray-400 mr-3 whitespace-nowrap">
                {license?.name}
              </div>
            </div>
          )}
          <div className="text-gray-400 font-normal text-xs">
            Updated on {updatedAt}
          </div>
        </div>
      </div>
    </div>
  );
};

//#e3b341