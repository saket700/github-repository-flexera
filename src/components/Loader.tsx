import React from "react";
import { LoadingIcon } from "../Icons";

export const Loader = () => (
    <div className="flex justify-start">
      <LoadingIcon />
      <span className="sr-only">Loading...</span>
    </div>
  );
  