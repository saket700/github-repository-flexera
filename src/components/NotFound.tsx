import React from "react";
interface INotFound {
  headingNotFound: string;
  descriptionNotFound: string;
}
export const NotFound: React.FC<INotFound> = ({
  headingNotFound,
  descriptionNotFound,
}) => (
  <div>
    <h3>{headingNotFound}</h3>
    <p>{descriptionNotFound}</p>
  </div>
);
