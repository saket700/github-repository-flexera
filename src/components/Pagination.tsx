import React from "react";
import classnames from "classnames";
import { DOTS, usePagination } from "../utils/hooks/usePagination";
import { LeftArrow, RightArrow } from "../Icons";

interface IPagination {
  onPageChange: (page: number) => void;
  totalCount: number;
  siblingCount: number;
  currentPage: number;
  pageSize: number;
}

export const Pagination: React.FC<IPagination> = ({
  onPageChange,
  totalCount,
  siblingCount = 1,
  currentPage,
  pageSize,
}) => {
  const paginationRange = usePagination({
    currentPage,
    totalCount,
    siblingCount,
    pageSize,
  }) as number[];

  if (!paginationRange || currentPage === 0 || paginationRange.length < 2) {
    return null;
  }

  const handlePageChange = (page: number) => {
    onPageChange(page);
  };

  const renderPaginationItem = (pageNumber: number, index: number) => (
    <li
      key={`${pageNumber}-${index}`}
      className={classnames(
        "flex items-center justify-center px-4 h-10 leading-tight text-gray-500 bg-white border border-gray-300 hover:bg-gray-100 hover:text-gray-700 dark:text-gray-400 dark:hover:bg-white-700 dark:hover:text-white",
        {
          "bg-white hover:bg-gray-100 hover:text-gray-700 dark:hover:bg-white-700 dark:hover:text-white":
            pageNumber === currentPage,
          "bg-gray-100 text-gray-700": pageNumber !== currentPage,
          "pointer-events-none opacity-50": pageNumber.toString() === DOTS,
        }
      )}
      onClick={() =>
        pageNumber.toString() === DOTS ? null : handlePageChange(pageNumber)
      }
    >
      {pageNumber.toString() === DOTS ? (
        <span className="dots">&#8230;</span>
      ) : (
        pageNumber
      )}
    </li>
  );
  let lastPage = paginationRange[paginationRange.length - 1];

  return (
    <nav aria-label="Page navigation">
      <ul className="flex items-center -space-x-px h-10 text-base">
        <li>
          <a
            href="#"
            onClick={(e) => {
              e.preventDefault();
              currentPage > 1 && handlePageChange(currentPage - 1);
            }}
            className={classnames(
              "flex items-center justify-center px-4 h-10 ms-0 leading-tight text-gray-500 bg-white border border-e-0 border-gray-300 rounded-s-lg hover:bg-gray-100 hover:text-gray-700  dark:hover:text-white",
              {
                "pointer-events-none opacity-50": currentPage === 1,
              }
            )}
          >
            <span className="sr-only">Previous</span>
            <LeftArrow />
          </a>
        </li>
        {paginationRange.map(renderPaginationItem)}
        <li>
          <a
            href="#"
            onClick={(e) => {
              e.preventDefault();
              handlePageChange(currentPage + 1);
            }}
            className={classnames(
              "flex items-center justify-center px-4 h-10 leading-tight text-gray-500 bg-white border border-gray-300 rounded-e-lg hover:bg-gray-100 dark:text-gray-400 dark:hover:text-white",
              {
                "pointer-events-none opacity-50": currentPage === lastPage,
              }
            )}
          >
            <span className="sr-only">Next</span>
            <RightArrow />
          </a>
        </li>
      </ul>
    </nav>
  );
};
