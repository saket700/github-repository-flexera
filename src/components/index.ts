export * from './Card';
export * from './CardLayout';
export * from './Loader';
export * from './NotFound';
export * from './Pagination';