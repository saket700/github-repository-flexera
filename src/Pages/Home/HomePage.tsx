import React, { useEffect, useState } from "react";
import { paginate, useGitHubRepos } from "../../utils";
import { Loader, Pagination } from "../../components";
import { RepoList } from "./components/RepoList";

const pageSize = 10;

export const HomePage = () => {
  const [currentPage, setCurrentPage] = useState(1);

  const { repos, loading, error, fetchGitHubRepos } = useGitHubRepos();
  const currentPosts = paginate(repos, currentPage, pageSize);
  useEffect(() => {
    const fetchData = async () => {
      try {
        await fetchGitHubRepos(currentPage, pageSize);
      } catch (error) {
        console.error("Error fetching GitHub repos:", error);
      }
    };

    fetchData();
  }, [currentPage]);

  return (
    <div className="m-auto max-w-[900px]">
      <label
        htmlFor="github_username"
        className="text-sm font-medium leading-6 text-gray-900 flex justify-center"
      >
        Github Explorer Repository
      </label>

      {loading && !error && (
        <div className="flex justify-center pt-10">
          <Loader />
        </div>
      )}
      {error ? (
        <div className="flex justify-center pt-10">
          <p className="text-red-600">
            Failed to fetch GitHub repositories. Please try again.
          </p>
        </div>
      ) : (
        <>
          <RepoList currentPosts={currentPosts?.items} />
          <div className="w-full flex justify-center">
            <Pagination
              currentPage={currentPage}
              totalCount={repos.totalCount}
              pageSize={pageSize}
              onPageChange={(page) => setCurrentPage(page)}
              siblingCount={0}
            />
          </div>
        </>
      )}
    </div>
  );
};

export default HomePage;
