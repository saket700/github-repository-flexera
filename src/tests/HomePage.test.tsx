import {
  fireEvent,
  render,
  waitFor,
  screen,
  act,
} from "@testing-library/react";
import HomePage from "../Pages/Home/HomePage";
import React from "react";
import { vi } from "vitest";
import { mockData } from "../mock-data/mockData";
import { Pagination } from "../components";
import renderer from "react-test-renderer";




describe("Snapshot testing suite", () => {
  it("Matches DOM Snapshot", () => {
    const domTree = renderer.create(<HomePage />).toJSON();
    expect(domTree).toMatchSnapshot();
  });
});

test("should render repositories from 1 to 10 on the first page", async () => {
  vi.mock("fetch", () => ({
    ok: true,
    json: () =>
      Promise.resolve({
        success: true,
        status: 200,
        data: mockData,
      }),
  }));
  
  render(<HomePage />);

  // Ensure that repositories from 1 to 10 are rendered
  await waitFor(() => {
    for (let i = 0; i < 10; i++) {
      const repoNameElement = screen.getByText(mockData[i].name);
      expect(repoNameElement).toBeInTheDocument();
    }
  });

  // Ensure that repositories from 11 onwards are not rendered
  for (let i = 10; i < mockData.length; i++) {
    const repoNameElement = screen.queryByText(mockData[i].name);
    expect(repoNameElement).toBeNull();
  }
});


beforeEach(() => {
  vi.resetAllMocks();
});

test("render error message on API failure", async () => {
  vi.spyOn(global, "fetch").mockRejectedValueOnce(new Error("Failed to fetch GitHub repositories"));

  await act(async () => {
    render(<HomePage />);
  });

  await waitFor(() => {
    const errorMessage = screen.getByText("Failed to fetch GitHub repositories. Please try again.");
    expect(errorMessage).toBeInTheDocument();
  });
});



test("should render the pagination buttons ", () => {
  const onPageChangeMock = vi.fn();
  const component = render(
    <Pagination
      onPageChange={onPageChangeMock}
      totalCount={10}
      siblingCount={1}
      currentPage={1}
      pageSize={5}
    />
  );

  const previousButton = screen.queryByText("Previous");
  const nextButton = screen.queryByText("Next");

  expect(previousButton).toBeInTheDocument();
  expect(nextButton).toBeInTheDocument();

  // Find and click the "Next" button
  const nextButtonClick = screen.getByText("Next");
  fireEvent.click(nextButtonClick);

  expect(onPageChangeMock).toHaveBeenCalledWith(2);
});
