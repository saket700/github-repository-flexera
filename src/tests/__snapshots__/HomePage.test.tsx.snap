// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot testing suite > Matches DOM Snapshot 1`] = `
<div
  className="m-auto max-w-[900px]"
>
  <label
    className="text-sm font-medium leading-6 text-gray-900 flex justify-center"
    htmlFor="github_username"
  >
    Github Explorer Repository
  </label>
  <div
    className="pt-8"
  />
  <div
    className="w-full flex justify-center"
  />
</div>
`;
